File: sandbox-python/scripts/live_egress_test.py
#!/usr/bin/env python3
"""Live egress test for brimble-sandbox Python SDK."""

from __future__ import annotations

import os
import sys
import time

from brimble_sandbox import Sandbox
from brimble_sandbox.enums import SandboxEgressMode

api_key = os.environ.get("BRIMBLE_SANDBOX_KEY")
if not api_key:
    print("Set BRIMBLE_SANDBOX_KEY to run this script.")
    sys.exit(1)

NETWORK_SWITCH_WAIT_S = 25
PROBE_CMD = "curl -s -o /dev/null -w '%{http_code}' --connect-timeout 5 http://1.1.1.1 || echo 000"


def probe_http(sandbox: Sandbox) -> str:
    result = sandbox.exec({"cmd": PROBE_CMD})
    stdout = str(result.get("stdout", "")).strip()
    digits = "".join(ch for ch in stdout if ch.isdigit())
    return digits[:3] or "000"


def assert_step(label: str, actual: str, expected: str) -> None:
    ok = actual == expected
    print(f"{'PASS' if ok else 'FAIL'} {label}: got {actual}, expected {expected}")
    if not ok:
        raise SystemExit(1)


client = Sandbox(api_key=api_key)

print("Creating sandbox with deny_all egress...")
sandbox = client.sandboxes.create(
    {
        "template": "node-22",
        "egress": {"mode": SandboxEgressMode.DENY_ALL},
    }
)
print(f"Sandbox id: {sandbox.id}")

try:
    sandbox.wait_until_ready(timeout_ms=180_000, poll_interval_ms=2_000)
    print("Sandbox ready.\n")

    assert_step("deny_all blocks outbound", probe_http(sandbox), "000")

    print("\nUpdating egress to restricted (allow 1.1.1.1)...")
    restricted = sandbox.update_egress(
        {"mode": SandboxEgressMode.RESTRICTED, "allow": ["1.1.1.1"]}
    )
    print(
        f"  egress.mode={restricted.get('egress', {}).get('mode')}, "
        f"network_updated={restricted.get('network_updated', False)}"
    )

    time.sleep(NETWORK_SWITCH_WAIT_S)
    assert_step("restricted allows 1.1.1.1", probe_http(sandbox), "301")

    print("\nUpdating egress to open...")
    open_result = sandbox.update_egress({"mode": SandboxEgressMode.OPEN})
    print(
        f"  egress.mode={open_result.get('egress', {}).get('mode')}, "
        f"network_updated={open_result.get('network_updated', False)}"
    )

    if open_result.get("network_updated"):
        time.sleep(NETWORK_SWITCH_WAIT_S)

    assert_step("open allows outbound", probe_http(sandbox), "301")
    print("\nAll egress SDK checks passed.")
finally:
    print("\nDestroying sandbox...")
    try:
        sandbox.destroy()
        print("Sandbox destroyed.")
    except Exception as error:  # noqa: BLE001
        print(f"Cleanup failed: {error}")

File: sandbox/scripts/live-egress-test.mjs
import { Sandbox, SandboxEgressMode } from '../dist/src/index.js';

const apiKey = process.env.BRIMBLE_SANDBOX_KEY;
if (!apiKey) {
  console.error('Set BRIMBLE_SANDBOX_KEY to run this script.');
  process.exit(1);
}

const networkSwitchWaitMs = 25_000;
const probeCmd =
  "curl -s -o /dev/null -w '%{http_code}' --connect-timeout 5 http://1.1.1.1 || echo 000";

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function probeHttp(sandbox) {
  const result = await sandbox.exec({ cmd: probeCmd });
  return result.stdout.trim().replace(/\D/g, '').slice(0, 3) || '000';
}

function assertStep(label, actual, expected) {
  const ok = actual === expected;
  console.log(`${ok ? 'PASS' : 'FAIL'} ${label}: got ${actual}, expected ${expected}`);
  if (!ok) {
    throw new Error(`${label} failed`);
  }
}

const client = new Sandbox({ apiKey, timeoutMs: 120_000 });

console.log('Creating sandbox with deny_all egress...');
const sandbox = await client.sandboxes.create({
  template: 'node-22',
  egress: { mode: SandboxEgressMode.DenyAll },
});

console.log(`Sandbox id: ${sandbox.id}`);

try {
  await sandbox.waitUntilReady({ timeoutMs: 180_000, pollIntervalMs: 2_000 });
  console.log('Sandbox ready.\n');

  const deniedAtBirth = await probeHttp(sandbox);
  assertStep('deny_all blocks outbound', deniedAtBirth, '000');

  console.log('\nUpdating egress to restricted (allow 1.1.1.1)...');
  const restricted = await sandbox.updateEgress({
    mode: SandboxEgressMode.Restricted,
    allow: ['1.1.1.1'],
  });
  console.log(`  egress.mode=${restricted.egress.mode}, network_updated=${restricted.network_updated ?? false}`);

  await sleep(networkSwitchWaitMs);

  const allowedProbe = await probeHttp(sandbox);
  assertStep('restricted allows 1.1.1.1', allowedProbe, '301');

  console.log('\nUpdating egress to open...');
  const open = await sandbox.updateEgress({ mode: SandboxEgressMode.Open });
  console.log(`  egress.mode=${open.egress.mode}, network_updated=${open.network_updated ?? false}`);

  if (open.network_updated) {
    await sleep(networkSwitchWaitMs);
  }

  const openProbe = await probeHttp(sandbox);
  assertStep('open allows outbound', openProbe, '301');

  console.log('\nAll egress SDK checks passed.');
} finally {
  console.log('\nDestroying sandbox...');
  try {
    await sandbox.destroy();
    console.log('Sandbox destroyed.');
  } catch (error) {
    console.warn('Cleanup failed:', error instanceof Error ? error.message : error);
  }
}
